contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+use proc_macro::{TokenStream, Diagnostic};
+use syn::{DataStruct, Fields, Data, Type, LitStr, DeriveInput, Ident};
+use proc_macro2::Span;
+
+use spanned::Spanned;
+
+type Result<T> = ::std::result::Result<T, Diagnostic>;
+
+#[derive(Debug)]
+struct DatabaseInvocation {
+    /// the name of the specified Database type that can be injected into request handlers
+    type_name: Ident,
+    /// The database name as passed in via #[database('database name')].
+    db_name: String,
+    /// The entire structure that the `database` attribute was called on.
+    structure: DataStruct,
+    /// The type inside the structure: struct MyDb(ThisType).
+    connection_type: Type,
+}
+
+const EXAMPLE: &str = "example: `struct MyDatabase(diesel::SqliteConnection);`";
+const ONLY_ON_STRUCTS_MSG: &str = "`database` attribute can only be used on structs";
+const ONLY_UNNAMED_FIELDS: &str = "`database` attribute can only be applied to structs with \
+    exactly one unnamed field";
+const NO_GENERIC_STRUCTS: &str = "`database` attribute cannot be applied to a struct with a \
+    generic type";
+
+const UNABLE_TO_APPLY: &str = "unable to apply the `database` attribute to the \
+    given struct";
+
+fn parse_invocation(attr: TokenStream, input: TokenStream) -> Result<DatabaseInvocation> {
+    let attr_stream2 = ::proc_macro2::TokenStream::from(attr);
+    let attr_span = attr_stream2.span();
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    if let Type::Path(type_path) = invocation.connection_type {
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
```

Comment:
I think this can be something like:

```rust
let pool_type = Ident::new(handler_type + "Pool", handler_type.span());
```

Note that `Span::call_site()` is not correct. We need to use the `Span` of the `handler_type` `Ident` since that's what we're generating it from.

Reply:
That's cleaner than what I have. Thanks for the note on the `Span` information. I'm still a bit fuzzy on the different types of `Span`s and need to take a deeper dive to truly understand them.

Reply:
Yeah, so `handler_type` is an Ident and you can't add a string onto that. I think `format!` is the best we can do here. At least that's the only option that's obvious to me.

Reply:
There is an `IdentExt::prepend` helper function... somewhere. I think in #680, adapted from the old codegen to work with the types in `syn`.

Reply:
For now, I think it would be most beneficial to wait until #680 lands to try to leverage that? Since the codegen crate and the contrib codegen crate are in different scopes, it could also make sense to just duplicate the `IdentExt` for use in both. For now, I'll not alter this, though.

Reply:
Oh. Looking at it again, prepend() does pretty much the same as what you already have after all, except for the Span.